### PR TITLE
Permit to read NTFS attributes data through read_random & list $Orphan files

### PR DESCRIPTION
**Pull request by EccoTheFlintstone**
Add the possibility to :
- list deleted files from the `$Orphans` pseudo directory (by being able to list files in the `$Orphans` dir)
- get data from NTFS attributes with an ID set to 0 (for instance some  $FILENAME or $STANDARD_INFO)

For the NTFS attributes, the `id` default value given to `read_random` is -1.
If it is not given, it is assumed that we want to read from the default attribute of the file ($DATA for NTFS, data stream for FAT...)
With this change, one can read data from other NTFS attributes and parse them

### CODE REVIEW

**Comment by joachimmetz** [comment]
My initial question, why overload this function (File_as_directory) ?
and not iterate the attributes as
https://github.com/py4n6/pytsk/wiki/Development#accessing-a-alternate-data-stream-ntfs

Best practices for an API is to provide a clear interface / semantics and overload a little as possible

**Comment by EccoTheFlintstone** [comment]
2 things in this PR :
- list deleted files from the `$Orphans `pseudo directory (by being able to list files in the $Orphans dir)

last year, sleuthkit changed the `$Orphans `pseudo dir from `TSK_FS_META_TYPE_DIR` to `TSK_FS_META_TYPE_VIRT_DIR` (see https://github.com/sleuthkit/sleuthkit/pull/874/files)
Without the change in this PR, the files in this pseudo directory (deleted files for instance) cannot be found 

The File_as_directory function is not overloaded, its behavior is only changed to permit listing the `$Orphans `directory and the files in it (it's probably a regression that started 1 year ago, see PR 874 above)

- get data from NTFS attributes with an ID set to 0 (for instance some `$FILENAME` or `$STANDARD_INFO`)

After testing again today, the `File_read_random` permits reading any attribute data of a file on a NTFS filesystem, not only an ADS. For instance, one can read the $FNAME data (type 0x30, generally ID == 2) of a file with this function.
Problem is, the ID of the attribute starts at 0. And generally (always?), the attribute with ID == 0 is the $STANDARD_INFORMATION one. Because of the check in File_read_random, one is able to read any attribute data EXCEPT the $STANDARD_INFORMATION one. In the case where ID == 0 is specified, `tsk_fs_file_read` is used instead of `tsk_fs_file_read_type`, and the read fails.

File: tsk3.c [hunk 1]
```diff
@@ -510,7 +510,7 @@ static Directory File_as_directory(File self) {
         RaiseError(EInvalidParameter, "Invalid parameter: self->info.");
         return NULL;
     }
-    if(self->info->meta == NULL || self->info->meta->type != TSK_FS_META_TYPE_DIR) {
+    if(self->info->meta == NULL || !(TSK_FS_IS_DIR_META(self->info->meta->type))) {
```
Comment (joachimmetz) [inline tsk3.c]: Would be checking for `TSK_FS_META_TYPE_VIRT_DIR` and `TSK_FS_META_TYPE_DIR` be sufficient?

This probably needs a libtsk version check as well.

File: tsk3.c [hunk 2]
```diff
@@ -484,7 +484,7 @@ static uint64_t File_read_random(File self, TSK_OFF_T offset,
     RaiseError(EInvalidParameter, "id parameter is invalid.");
     return 0;
   };
-  if(id > 0) {
+  if(id >= 0) {
```
Comment (joachimmetz) [inline tsk3.c]: Doesn't this change means the else never returns successfully, since it will hit line 493 an you would need to determine the id of the default (nameless) data attribute?

From http://www.sleuthkit.org/sleuthkit/docs/api-docs/4.5/group__fslib.html#ga603ea8f80f1ffbf6be888aeaa508fc81

```
a_id | The id of attribute to load (use 0 and set a_flags if you do not care)
```

http://www.sleuthkit.org/sleuthkit/docs/api-docs/4.5/group__fslib.html#ga54806cf281c7879dcaf7042ff267f447

File: tsk3.c (hunk 1 above)
Comment (EccoTheFlintstone) [reply tsk3.c]: The `TSK_FS_IS_DIR_META` macro does exactly this :
https://github.com/sleuthkit/sleuthkit/blob/develop/tsk/fs/tsk_fs.h#L403

Comment (joachimmetz) [reply tsk3.c]: Ack but it was introduced in a specific version of TSK https://github.com/sleuthkit/sleuthkit/blame/develop/tsk/fs/tsk_fs.h#L403

please add version guards

File: tsk3.c (hunk 2 above)
Comment (EccoTheFlintstone) [reply tsk3.c]: The else is taken when you do not specify an ID, as a default ID of -1 is specified in the python part (don't remember where, maybe in the generated `pytsk3.c`). This is the case when you specify `offset` and `length` only (for instance https://github.com/py4n6/pytsk/blob/886a4bab2979f780e26680b32aec8a50350fe2d2/examples/icat.py#L61)

If you do not care about ID, you can set 0 (or whatever ID, as it's not checked at all in this case) and set the `TSK_FS_FILE_READ_FLAG_NOID` flag (https://github.com/sleuthkit/sleuthkit/blob/9e4bc4b5c0d34dcb266191558337be1e199af3da/tsk/fs/tsk_fs.h#L696)

In `tsk_fs_file_read_type` (https://github.com/sleuthkit/sleuthkit/blob/9e4bc4b5c0d34dcb266191558337be1e199af3da/tsk/fs/fs_file.c#L474), the `fs_attr` value is retrieved using the `type` (and `id` if the `TSK_FS_FILE_READ_FLAG_NOID` flag is not specified)
https://github.com/sleuthkit/sleuthkit/blob/9e4bc4b5c0d34dcb266191558337be1e199af3da/tsk/fs/fs_file.c#L499  
Adn then the content of the attribute is read with `tsk_fs_attr_read`.

This is also what happens in `tsk_fs_file_read` where the default attribute is retrieved with `tsk_fs_file_attr_get` and then also read with `tsk_fs_attr_read`


To sum things up,  if we didn't need the `id` parameter, we should set the `TSK_FS_FILE_READ_FLAG_NOID` flag to retrieve the content of a an attribute (specified by `type`) without caring about its ID.
But as we could need to retrieve an `id` with value 0, this change is needed and it should be the caller responsibility to set the `TSK_FS_FILE_READ_FLAG_NOID` flag if he wants data from a specific `type` but without specifying the ID. (as is done in libtsk actually)

**Comment by joachimmetz** [comment]
thx, for adding the version guard. When time permits, I'll have a closer look at the tsk source and the id; just to rule out unwanted side effects.

**Comment by joachimmetz** [comment]
https://github.com/sleuthkit/sleuthkit/blob/9e4bc4b5c0d34dcb266191558337be1e199af3da/tsk/fs/fs_file.c#L522

```
/**
 * \ingroup fslib
 * Read the contents of a specific attribute of a file using a typical read() type interface.
 * 0s are returned for missing runs of files. 
 * 
 * @param a_fs_file The inode structure of the file to read.
 * @param a_offset The byte offset to start reading from.
 * @param a_buf The buffer to read the data into.
 * @param a_len The number of bytes to read from the file.
 * @param a_flags Flags to use while reading
 * @returns The number of bytes read or -1 on error (incl if offset is past EOF).
 */
ssize_t
tsk_fs_file_read(TSK_FS_FILE * a_fs_file,
    TSK_OFF_T a_offset, char *a_buf, size_t a_len,
    TSK_FS_FILE_READ_FLAG_ENUM a_flags)
```

https://github.com/sleuthkit/sleuthkit/blob/9e4bc4b5c0d34dcb266191558337be1e199af3da/tsk/fs/fs_file.c#L474

```
/**
* \ingroup fslib
 * Read the contents of a specific attribute of a file using a typical read() type interface and be
 * able specify a specific attribute to read (applies only to file systems with multiple attributes
 * per file, such as NTFS).  0s are returned for missing runs of files. 
 * 
 * @param a_fs_file The file to read from
 * @param a_type The type of attribute to load
 * @param a_id The id of attribute to load (use 0 and set a_flags if you do not care)
 * @param a_offset The byte offset to start reading from.
 * @param a_buf The buffer to read the data into.
 * @param a_len The number of bytes to read from the file.
 * @param a_flags Flags to use while reading
 * @returns The number of bytes read or -1 on error (incl if offset is past EOF).
 */
ssize_t
tsk_fs_file_read_type(TSK_FS_FILE * a_fs_file,
    TSK_FS_ATTR_TYPE_ENUM a_type, uint16_t a_id, TSK_OFF_T a_offset,
    char *a_buf, size_t a_len, TSK_FS_FILE_READ_FLAG_ENUM a_flags)
```

File: tsk3.c (hunk 2 above)
Comment (joachimmetz) [reply tsk3.c]: I'm moving this to a separate change.

**Comment by joachimmetz** [comment]
Generated pytsk3.c
```
static PyObject *pyFile_read_random(pyFile *self, PyObject *args, PyObject *kwds){
    PyObject *returned_result = NULL;
    PyObject *Py_result = NULL;
```

Sets id to `-1`,

**Comment by joachimmetz** [comment]
Addition changes https://github.com/py4n6/pytsk/commit/7f57f6870bc7436ed66f194cdccdde87f52ff81c
